### ISSUE TITLE
Cell definition with contour coordinates

### ISSUE THREAD

**Issue by MarAlder** [DLR-SL/CPACS#656]
@ewerthen provides a proposal to extend the current cell definition by contour variables:

old:
![grafik](https://user-images.githubusercontent.com/43143741/94429648-ebed9100-0192-11eb-8077-7689610002dc.png)

new:
![grafik](https://user-images.githubusercontent.com/43143741/94429712-0889c900-0193-11eb-82f0-5b8bf95a917c.png)

Implemented in: https://github.com/DLR-SL/CPACS/tree/cellCoordinates

**Comment by MarAlder** [comment]
As a question to @ewerthen and in the round: `xsi1` and `xsi2` is defined in CPACS as coordinates on the wing cord-plane. The lofting algorithm, however, is not specified in CPACS. Thus, `innerContourCoordinate` and `outerContourCoordinate` would also not be defined :confused:. Of course we hope that most users will use [TiGL](https://github.com/DLR-SC/tigl/) as a geometry library, with which the contour coordinates are defined via the implemented lofting algorithm. But shouldn't the data exchange format be independent of any implementation, whether TiGL or other software?

Or can we use the documentation in CPACS to specify which lofting algorithm the new parameters refer to in order to avoid this conflict?

Why exactly are the coordinates not sufficiently defined via `xsi1` and `xsi2`?

**Comment by hardt-fa-flb** [comment]
We use the wing cell to define the material distribution of the wing / blade. Most wind turbine blade have leading and trailing edge reinforcements (see image).

![grafik](https://user-images.githubusercontent.com/41366119/94740406-e64ea180-0372-11eb-9af8-8c31f4e0ae3a.png)

The cells for these reinforcements cannot be defined with `xsi1` and `xsi2` because there the profile contour runs almost perpendicular to the chord. That is the reason why we need the definition with contour coordinates.

Another reason for the definition of the cells with contour coordinates is that the generation of the shapes of the cells can be significantly accelerated in our optimization (cell border is defined with isocurves, @rainman110 ).

According to @rainman110 , the loft contour coordinate in TIGL are defined as follows: It runs from -1 on the trailing edge on the lower side through 0 on the leading edge to 1 on the trailing edge on the upper side. I think a similarly normalized definition is also suitable for CPACS.

![grafik](https://user-images.githubusercontent.com/41366119/94740821-87d5f300-0373-11eb-893b-6c8a006eed23.png)

**Comment by ewerthen** [comment]
Like @hardt-fa-flb wrote "The cells for these reinforcements cannot be defined with xsi1 and xsi2 because there the profile contour runs almost perpendicular to the chord" e.g. at the trailing edge which would create a cell where xsi1 and xsi2 have nearly the same values and the Cell generation fails.

**Comment by MarAlder** [comment]
Thanks for the detailed explanations! Any feedback from the community?

**Comment by hardt-fa-flb** [comment]
After consultation with @joergbrech we propose to replace the inner and outer contour coordinate with just one contour coordinate. With just one contour coordinate, the chordwise border can be set up as an isocurve in TIGL, which has advantages performance.

![grafik](https://user-images.githubusercontent.com/41366119/101450409-cca97780-392a-11eb-932c-5dd6d12349f6.png)

[cpacs_schema.zip](https://github.com/DLR-SL/CPACS/files/5657385/cpacs_schema.zip)

**Comment by joergbrech** [comment]
@hardt-fa-flb, would it make sense to add this option also for the spanwise cell positioning? I think this could help to speed up the calculation of cell geometries in TiGL as well.

**Comment by hardt-fa-flb** [comment]
Hi @joergbrech, is the spanwise positioning with eta on the chordface equivalent (or at least very similar) to the definition of the spanwise positioning with a contour coordinate on the shell? Wouldn't it be possible there to use the existing definition and generate the isocurve if eta1 = eta2?

**Comment by joergbrech** [comment]
Yes, in *almost* all cases I would expect that eta1=eta2 yields an isocurve on the wing shell. But there are exceptions, e.g. if the wing is defined using guide curves.  So TiGL would have to check if the border coincides with an isocurve, this is not a trivial task and the check alone will likely decrease performance. If a contour coordinate is provided the definition is unambiguous and the check can be avoided.

**Comment by hardt-fa-flb** [comment]
We have no problem with the span-wide positioning as a contour coordinate. But don't you get problems in a spanwise direction anyway, since the wing is divided into segments and the spanwise contour coordinates lie on several segments?

**Comment by joergbrech** [comment]
You are right, but I don't see any difference to the chordwise positioning: If a wing has guide curves, it can also happen that the contour coordinates lie on several faces/surfaces and in TiGL we would have to account for this special case.

Don't get me wrong, I don't want to insist on spanwise ContourCoordinates, I can perfectly live without it. My argument is simply that I currently see no reason in discriminating between a spanwise and a chordwise ContourCoordinate. From a CPACS definition point of view it seems more logical to me to use the same definition in both directions. 

Also, I have the hope that the performance can be increased even further for those special cases where a wing cell does not intersect face/surface boundaries on the wing shell.

**Comment by hardt-fa-flb** [comment]
I totally agree with you. For our application, the definition with a contour coordinate in spanwise direction is completely okay. If it a lot faster, I'll just create a new scheme. As we said, it's primarily about performance.

**Comment by hardt-fa-flb** [comment]
![grafik](https://user-images.githubusercontent.com/41366119/101476886-b7464480-394e-11eb-86ad-1ffe7811f349.png)

![grafik](https://user-images.githubusercontent.com/41366119/101476919-c3ca9d00-394e-11eb-85e8-f4b16a7047ad.png)

[cpacs_schema.zip](https://github.com/DLR-SL/CPACS/files/5658882/cpacs_schema.zip)

**Comment by ewerthen** [comment]
Perfect, thank you @joergbrech and @hardt-fa-flb. @MarAlder can you integrate it in CPACS 3.4?

**Comment by ewerthen** [comment]
I spoke to Marco, he has given his blessing :-)

**Comment by joergbrech** [comment]
I implemented this in TiGL in this PR: https://github.com/DLR-SC/tigl/pull/769, and now, in hindsight, have two minor comments:

 - I suggest renaming `spanwiseContourCoordinate` and `chordwiseContourCoordinate` simply to `contourCoordinate`. Since it is nested in the respective `cellPositioningXXXType`, this would still be unambiguous and less verbose.
 - With this proposal it is possible to mix all four border definitions: Some can be defined relatively to the chordface, e.g. by defining eta/xsi coordinates or rib/spar references, and others can be defined relatively to the skin surface by prescribing contour coordinates. But mixing the border definitions is actually a bit problematic for TiGL, because of possible robustness issues. Therefore, TiGL doesn't support this. In other words, **if one border is defined via contour coordinates, all borders must be defined using contour coordinates**. Now the question is, wether we want to enforce this restriction on CPACS level as well.

Regarding the second point, I don't have a strong opinion for either case. What do you think @MarAlder?

**Comment by ewerthen** [comment]
I would recommend to follow the proposal of point 2!

**Comment by MarAlder** [comment]
@ewerthen, what do you mean with following point 2?

@joergbrech:
- renaming: I think the type should not provide semantic meaning. However, since the element names also indicate that only one of the two contourCoordinates makes sense, I would support to rename the coordinate elements to `contourCoordinate` and additionaly mention it in the documentation to support new users in coming to this conclusion right away. But since there are actually two independent coordinate directions I also wouldn't mind to follow @hardt-fa-flb proposal.  So, I am leaving the decision to the users at this point. @ewerthen , @hardt-fa-flb ?
- coordinate definitions: Hmm, do you have a solution in mind? Because I think it would require grouping the `positioning...` nodes into a choide element one level above, which could look quite like an implementation-related workaround. Do you think that providing a TiGL error message like `mixed contour coordinate definitions not supported by TiGL` or so would cast a bad light on TiGL? 

A little XSD remark before finishing the TiGL implementation: The sequence element above the first choice element is superfluous: 
[cpacs_schema.zip](https://github.com/DLR-SL/CPACS/files/5894612/cpacs_schema.zip)

**Comment by joergbrech** [comment]
> Do you think that providing a TiGL error message like mixed contour coordinate definitions not supported by TiGL or so would cast a bad light on TiGL?

No, that is absolutely fine with me. I just wanted to open the discussion here before this proposal becomes part of the standard and we cannot change the CPACS definition afterwards. Adding a choice element at a higher level is probably more intrusive, so lets keep it the way it is.

**Comment by ewerthen** [comment]
Perfect

**Comment by hardt-fa-flb** [comment]
For me the renaming is absolutely fine. Defining all borders either with eta/xsi-coordinates or contour-cordinates is also no problem. In fact, we already use the border with contour coordinates on all four sides, as this has increased the performance very much.

**Comment by MarAlder** [comment]
Voilà (26b8d9b):

![grafik](https://user-images.githubusercontent.com/43143741/106439170-afe77700-6477-11eb-945d-66490a83308e.png)

```XML
<xsd:documentation>Chordwise contour coordinate as chordwise border. 0 equals LE, 1 equals TE.</xsd:documentation>
```

and 

```XML
<xsd:documentation>Spanwise contour coordinate as spanwise border. 0 equals root, 1 equals tip.</xsd:documentation>
```
